include/power_grid_model/math_solver/iterative_pf_solver.hpp

```diff
@@ -0,0 +1,170 @@
+// SPDX-FileCopyrightText: 2022 Contributors to the Power Grid Model project <dynamic.grid.calculation@alliander.com>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_MATH_SOLVER_ITERATIVE_PF_SOLVER_HPP
+#define POWER_GRID_MODEL_MATH_SOLVER_ITERATIVE_PF_SOLVER_HPP
+
+/*
+ * Class to house common functions of newton raphson and iterative current method
+ */
+
+// Check if all includes needed
+#include "../calculation_parameters.hpp"
+#include "../exception.hpp"
+#include "../power_grid_model.hpp"
+#include "../three_phase_tensor.hpp"
+#include "../timer.hpp"
+#include "y_bus.hpp"
+
+namespace power_grid_model {
+
+// hide implementation in inside namespace
+namespace math_model_impl {
+
+// solver
+template <bool sym, typename DerivedSolver>
+class IterativePFSolver {
+   public:
+    MathOutput<sym> run_power_flow(YBus<sym> const& y_bus, PowerFlowInput<sym> const& input, double err_tol,
+                                   Idx max_iter, CalculationInfo& calculation_info) {
+        // get derived pointer for derived solver class
+        DerivedSolver& derived_solver = static_cast<DerivedSolver&>(*this);
+        IdxVector const& source_bus_indptr = *source_bus_indptr_;
+        std::vector<double> const& phase_shift = *phase_shift_;
+
+        // prepare
+        MathOutput<sym> output;
+        output.u.resize(n_bus_);
+        double max_dev = std::numeric_limits<double>::max();
+
+        Timer main_timer(calculation_info, 2220, "Math solver");
+
+        // initialize
+        Timer sub_timer(calculation_info, 2221, "Initialize calculation");
+        // average u_ref of all sources
+        DoubleComplex const u_ref = [&]() {
+            DoubleComplex sum_u_ref = 0.0;
+            for (Idx bus = 0; bus != n_bus_; ++bus) {
+                for (Idx source = source_bus_indptr[bus]; source != source_bus_indptr[bus + 1]; ++source) {
+                    sum_u_ref += input.source[source] * std::exp(1.0i * -phase_shift[bus]);  // offset phase shift
+                }
+            }
+            return sum_u_ref / (double)input.source.size();
+        }();
+
+        // assign u_ref as flat start
+        for (Idx i = 0; i != n_bus_; ++i) {
+            // consider phase shift
+            output.u[i] = ComplexValue<sym>{u_ref * std::exp(1.0i * phase_shift[i])};
+        }
+        // initialize unknown for polar complex for NR, nothing for iterative current
+        derived_solver.initialize_unknown_polar(output);
+        sub_timer.stop();
+        // Not used in NR, initialize non changing lhs for iterative current
+        derived_solver.initialize_matrix(y_bus);
+
+        // start calculation
+        // iteration
+        Idx num_iter = 0;
+        while (max_dev > err_tol) {
+            if (num_iter++ == max_iter) {
+                throw IterationDiverge{max_iter, max_dev, err_tol};
+            }
+            sub_timer = Timer(calculation_info, 2222, "Calculate jacobian and rhs");
```

Comment:
same goes here. 

Put each calculation step into a separate block with timer defined inside. This pattern is called RAII.